src/sqlancer/tidb/oracle/TiDBCERTOracle.java

```diff
@@ -0,0 +1,285 @@
+import sqlancer.common.query.SQLancerResultSet;
+import sqlancer.tidb.TiDBErrors;
+import sqlancer.tidb.TiDBExpressionGenerator;
+import sqlancer.tidb.TiDBProvider.TiDBGlobalState;
+import sqlancer.tidb.TiDBSchema.TiDBColumn;
+import sqlancer.tidb.TiDBSchema.TiDBDataType;
+import sqlancer.tidb.TiDBSchema.TiDBTables;
+import sqlancer.tidb.ast.TiDBBinaryLogicalOperation;
+import sqlancer.tidb.ast.TiDBBinaryLogicalOperation.TiDBBinaryLogicalOperator;
+import sqlancer.tidb.ast.TiDBColumnReference;
+import sqlancer.tidb.ast.TiDBExpression;
+import sqlancer.tidb.ast.TiDBJoin;
+import sqlancer.tidb.ast.TiDBJoin.JoinType;
+import sqlancer.tidb.ast.TiDBSelect;
+import sqlancer.tidb.ast.TiDBTableReference;
+import sqlancer.tidb.visitor.TiDBVisitor;
+
+public class TiDBCERTOracle implements TestOracle<TiDBGlobalState> {
+    private final TiDBGlobalState state;
+    private TiDBExpressionGenerator gen;
+    private final ExpectedErrors errors = new ExpectedErrors();
+    private TiDBSelect select;
+    private List<String> queryPlan1Sequences;
+    private List<String> queryPlan2Sequences;
+
+    public TiDBCERTOracle(TiDBGlobalState globalState) {
+        state = globalState;
+        TiDBErrors.addExpressionErrors(errors);
+    }
+
+    @Override
+    public void check() throws SQLException {
+        queryPlan1Sequences = new ArrayList<>();
+        queryPlan2Sequences = new ArrayList<>();
+
+        // Randomly generate a query
+        TiDBTables tables = state.getSchema().getRandomTableNonEmptyTables();
+        gen = new TiDBExpressionGenerator(state).setColumns(tables.getColumns());
+        select = new TiDBSelect();
+
+        List<TiDBExpression> fetchColumns = new ArrayList<>();
+        fetchColumns.addAll(Randomly.nonEmptySubset(tables.getColumns()).stream().map(c -> new TiDBColumnReference(c))
+                .collect(Collectors.toList()));
+        select.setFetchColumns(fetchColumns);
+
+        List<TiDBExpression> tableList = tables.getTables().stream().map(t -> new TiDBTableReference(t))
+                .collect(Collectors.toList());
+        List<TiDBExpression> joins = TiDBJoin.getJoinsWithoutNature(tableList, state);
+        select.setJoinList(joins);
+        select.setFromList(tableList);
+        if (Randomly.getBoolean()) {
+            select.setWhereClause(gen.generateExpression());
+        }
+        if (Randomly.getBooleanWithRatherLowProbability()) {
+            select.setOrderByExpressions(gen.generateOrderBys());
+        }
+        if (Randomly.getBoolean()) {
+            select.setGroupByExpressions(select.getFetchColumns());
+            if (Randomly.getBoolean()) {
+                select.setHavingClause(gen.generateHavingClause());
+            }
+        }
+        if (Randomly.getBoolean()) {
+            select.setLimitClause(gen.generateExpression());
+        }
+        if (Randomly.getBoolean()) {
+            select.setOffsetClause(gen.generateExpression());
+        }
```

Comment:
Could we extract this to a method that returns a random `SELECT` statement? Perhaps this would be something that the other test oracles might be able to use?

Reply:
Actually I have some customized code to generate satisfied `SELECT` statements, such as no natural join

Reply:
I see. Then, perhaps, we can just extract a method but leave the other parts as is?

Reply:
It will absolutely make the code more complicated, as we do not have clear application scenarios that how other oracles use it while we create a complicated architecture. Therefore, I prefer to leave it as it is to make the code simple.